src/test/java/com/solvd/AbstractTest.java

```diff
@@ -32,16 +44,46 @@ public void setUp(String browser) throws MalformedURLException {
         } else {
             throw new IllegalArgumentException("Invalid browser parameter");
         }
+        browserName = browser;
         driver.manage().window().maximize();
         driver.manage().timeouts().pageLoadTimeout(Duration.ofSeconds(10));
         driverThreadLocal.set(driver);
     }
 
-    @AfterMethod
+    @AfterClass
     public void tearDown() {
         WebDriver driver = driverThreadLocal.get();
         if (driver != null) {
             driver.quit();
         }
     }
+
+    @AfterMethod
+    public void tearDown2(ITestResult testResult) {
+        if (testResult.getStatus() == ITestResult.FAILURE) {
+            takeScreenshotIfFailed(driverThreadLocal.get(), testResult.getMethod().getMethodName());
+        }
+    }
+
+    private void takeScreenshotIfFailed(WebDriver driver, String methodName) {
+        File screenshot = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
+        String screenshotPath = "screenshots/" + methodName + "_" + getBrowserName() + "_" + getCurrentDateTime() + ".png";
```

Comment:
You can use String.format for screenshotPath, as with every "+" you create a new string object. So if many tests fail you may run out of memory.

Reply:
good note. I fixed it. thank you)